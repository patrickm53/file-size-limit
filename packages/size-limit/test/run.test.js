let { join } = require('path')

let run = require('../run')

const ROOT = join(__dirname, '..', '..', '..')

function createProcess (fixture, args = []) {
  let history = {
    exitCode: 0,
    stdout: '',
    stderr: ''
  }
  let process = {
    argv: ['node', 'size-limit', ...args],
    cwd () {
      if (fixture.includes('/')) {
        return fixture
      } else {
        return join(__dirname, 'fixtures', fixture)
      }
    },
    exit (code) {
      history.exitCode = code
    },
    stdout: {
      write (str) {
        history.stdout += str.replace(ROOT, '')
      }
    },
    stderr: {
      write (str) {
        history.stderr += str.replace(ROOT, '')
      }
    }
  }
  return [process, history]
}

async function check (fixture, args) {
  let [process, history] = createProcess(fixture, args)
  await run(process)
  expect(history.stderr).toEqual('')
  expect(history.exitCode).toEqual(0)
  return history.stdout
}

async function error (fixture, args) {
  let [process, history] = createProcess(fixture, args)
  await run(process)
  expect(history.stdout).toEqual('')
  expect(history.exitCode).toEqual(1)
  return history.stderr
}

it('shows version', async () => {
  let out = await check('file', ['--version'])
  expect(out).toMatch(/^size-limit \d+.\d+.\d+\n$/m)
})

it('shows help', async () => {
  let out = await check('file', ['--help'])
  expect(out).toContain('Core options:')
  expect(out).not.toContain('Webpack options:')
})

it('shows webpack-related help', async () => {
  let out = await check('webpack', ['--help'])
  expect(out).toContain('Core options:')
  expect(out).toContain('Webpack options:')
})

it('uses dependencies to detect modules', async () => {
  let out = await check('dependencies', ['--help'])
  expect(out).toContain('Webpack options:')
})

it('shows error on missed package.json', async () => {
  let stderr = await error('/')
  expect(stderr).toMatchSnapshot()
})

it('shows syntax errors in package.json', async () => {
  let stderr = await error('package-syntax')
  expect(stderr).toMatchSnapshot()
})

it('shows own error in JSON format', async () => {
  let [process, history] = createProcess('/', ['--json'])
  await run(process)
  expect(history.exitCode).toEqual(1)
  expect(history.stderr).toEqual('')
  expect(history.stdout).toMatchSnapshot()
})

it('shows external error in JSON format', async () => {
  let [process, history] = createProcess('package-syntax', ['--json'])
  await run(process)
  expect(history.exitCode).toEqual(1)
  expect(history.stderr).toEqual('')
  expect(history.stdout).toMatchSnapshot()
})
