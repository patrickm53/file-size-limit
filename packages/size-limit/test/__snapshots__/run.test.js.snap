// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows error on missed package.json 1`] = `
"[41m[30m ERROR [39m[49m [31mSize Limit didâ€™t find [33mpackage.json[31m.[39m
[31m        Create npm package and run Size Limit there.[39m
"
`;

exports[`shows migration guide for npm users 1`] = `
"[31mInstall Size Limit preset depends on type of the project[39m

For application, where you send JS bundle directly to users
  [33mnpm install --save-dev @size-limit/preset-app[39m
For frameworks, components and big libraries
  [33mnpm install --save-dev @size-limit/preset-big-lib[39m
For small (< 10 KB) libraries
  [33mnpm install --save-dev @size-limit/preset-small-lib[39m
Check out docs for more comlicated cases
  [33mhttps://github.com/ai/size-limit/[39m
"
`;

exports[`shows migration guide for yarn users 1`] = `
"[31mInstall Size Limit preset depends on type of the project[39m

For application, where you send JS bundle directly to users
  [33myarn add --dev @size-limit/preset-app[39m
For frameworks, components and big libraries
  [33myarn add --dev @size-limit/preset-big-lib[39m
For small (< 10 KB) libraries
  [33myarn add --dev @size-limit/preset-small-lib[39m
Check out docs for more comlicated cases
  [33mhttps://github.com/ai/size-limit/[39m

You can remove size-limit dependency: [33myarn remove size-limit[39m
"
`;

exports[`throws on --why argument without webpack 1`] = `
"[41m[30m ERROR [39m[49m [31mArgument [33m--why[31m works only with [33m@size-limit/webpack[31m module.[39m
[31m        You can add Bundle Analyzer to you own bundler.[39m
"
`;

exports[`throws on other webpack arguments without webpack 1`] = `
"[41m[30m ERROR [39m[49m [31mArgument [33m--webpack-config[31m works only with [33m@size-limit/webpack[31m module[39m
"
`;

exports[`throws on unknown argument 1`] = `
"[41m[30m ERROR [39m[49m [31mUknown argument [33m--unknown[31m.[39m
[31m        Check command for typo and read docs.[39m
"
`;
